### PR TITLE
[eroprofile] Add album downloader

### PR DESCRIPTION
**Pull request by jhwgh1968**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [x] New extractor

---

### Description of your *pull request* and other information

This adds an album extractor to adult website EroProfile.

I don't think it needs to handle authentication, as each video should handle logging in on its own.

### CODE REVIEW

File: yt_dlp/extractor/eroprofile.py [hunk 1]
```diff
@@ -90,3 +90,48 @@ def _real_extract(self, url):
             'title': title,
             'age_limit': 18,
         })
+
+
+class EroProfileAlbumIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?eroprofile\.com/m/videos/album/(?P<id>[^/]+)'
+    _NETRC_MACHINE = 'eroprofile'
```
Comment (pukkandan) [inline yt_dlp/extractor/eroprofile.py:97]:
```suggestion
```
since you aren't logging in

File: yt_dlp/extractor/eroprofile.py [hunk 2]
```diff
@@ -90,3 +90,48 @@ def _real_extract(self, url):
...
+            page_url = url + "?pnum={}".format(n)
+            playlist_pages.append(self._download_webpage(
+                page_url,
+                playlist_id,
+                note='Downloading playlist page {} of {}'.format(n, max_page)))
+
+        video_links = []
+        for webpage in playlist_pages:
+            video_links += ["https://www.eroprofile.com" + uri
+                            for uri in re.findall(r'href=".*?(/m/videos/view/[^"]+)"', webpage)]
```
Comment (pukkandan) [inline yt_dlp/extractor/eroprofile.py:133]: Use `InAdvancePagedList` if the number of items in a page is known in advance, else yield the videos

See https://github.com/yt-dlp/yt-dlp/pull/657#discussion_r685516029

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by jhwgh1968** [comment]
I believe I have fixed it, @pukkandan.

There is one odd thing with this: the output message order. Is this okay?
```
[debug] [EroProfile:album] Extracting URL: https://www.eroprofile.com/m/videos/album/....
[EroProfile:album] ...: Downloading playlist page 1
[download] Downloading playlist: ...
[EroProfile:album] ...: Downloading playlist page 2
[EroProfile:album] ...: Downloading playlist page 3
```

File: yt_dlp/extractor/eroprofile.py [hunk 3]
```diff
@@ -90,3 +90,48 @@ def _real_extract(self, url):
...
+        first_page = self._download_webpage(url, playlist_id, note='Downloading playlist page 1')
+
+        playlist_title = self._search_regex(
+            r'<title>Album: (.*) - EroProfile</title>', first_page, 'playlist_title')
+
+        entries = self._entries(url, playlist_title, first_page)
+
+        return self.playlist_result(entries,
+                                    playlist_id,
+                                    playlist_title)
```
Comment (pukkandan) [inline yt_dlp/extractor/eroprofile.py:137]:
```suggestion
    def _extract_from_page(self, page):
        for url in re.findall(r'href=".*?(/m/videos/view/[^"]+)"', page):
            yield self.url_result(f'https://www.eroprofile.com{url}', EroProfileIE.ie_key())

    def _entries(self, playlist_id, first_page):
        yield from self._extract_from_page(first_page)

        page_urls = re.findall(rf'href=".*?(/m/videos/album/{playlist_id}\?pnum=(\d+))"', first_page)

        for url, n in page_urls[1:]:
            yield from self._extract_from_page(self._download_webpage(
                f'https://www.eroprofile.com{url}',
                playlist_id, note=f'Downloading playlist page {int(n) - 1}'))

    def _real_extract(self, url):
        playlist_id = self._match_id(url)
        first_page = self._download_webpage(url, playlist_id, note='Downloading playlist')
        playlist_title = self._search_regex(
            r'<title>Album: (.*) - EroProfile</title>', first_page, 'playlist_title')

        return self.playlist_result(self._entries(playlist_id, first_page), playlist_id, playlist_title)
```

**Comment by jhwgh1968** [comment]
@pukkandan it seems that the yield refactor caused it to skip over pages. Hopefully I will have a second PR to fix this soon.

**Pull request merged**